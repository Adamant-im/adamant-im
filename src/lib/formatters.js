import { EPOCH } from './constants'

function getTime (date) {
  const hours = date.getHours()
  let time = ''
  if (hours < 10) {
    time = '0' + hours
  } else {
    time = '' + hours
  }
  time = time + ':'
  const minutes = date.getMinutes()
  if (minutes < 10) {
    time = time + '0' + minutes
  } else {
    time = time + '' + minutes
  }
  return time
}

function formatDate (timestamp) {
  timestamp = parseInt(timestamp)
  // That's for the ADM timestamps, which use EPOCH as a base.
  // Other cryptos use normal timestamps
  if (timestamp < EPOCH) {
    timestamp = timestamp * 1000 + EPOCH
  }

  const startToday = new Date()
  startToday.setHours(0, 0, 0, 0)

  const date = new Date(timestamp)
  if (date.getTime() > startToday.getTime()) {
    return this.$t('chats.date_today') + ', ' + getTime(date)
  }

  const startYesterday = new Date(startToday.getTime() - 86400000)
  if (date.getTime() > startYesterday.getTime()) {
    return this.$t('chats.date_yesterday') + ', ' + getTime(date)
  }

  let options = { weekday: 'short' }
  if ((Date.now() - timestamp) > (4 * 3600 * 24 * 1000)) {
    options = { day: 'numeric', month: 'short' }
  }

  if (startToday.getFullYear() !== date.getFullYear()) {
    options.year = 'numeric'
  }

  return date.toLocaleDateString(this.$t('region'), options) + ', ' + getTime(date)
}

const plugin = {
  install (app) {
    app.config.globalProperties.$formatDate = formatDate
  }
}

export default plugin
