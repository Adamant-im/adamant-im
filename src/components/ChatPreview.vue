<template>
  <v-list-tile
    class="chat-preview"
    @click="$emit('click')"
  >
    <v-list-tile-avatar>
      <icon v-if="readOnly" class="adm-icon"><adm-fill-icon/></icon>
      <chat-avatar v-else :size="40" :user-id="partnerId" use-public-key/>

      <v-badge overlap color="primary">
        <span v-if="numOfNewMessages > 0" slot="badge">
          {{ numOfNewMessages > 99 ? '99+' : numOfNewMessages }}
        </span>
      </v-badge>
    </v-list-tile-avatar>

    <v-list-tile-content>
      <v-list-tile-title v-text="readOnly ? $t(partnerName) : partnerName"></v-list-tile-title>

      <!-- Transaction -->
      <template v-if="lastTransaction">
        <v-list-tile-sub-title>
          <v-icon size="15">{{ statusIcon }}</v-icon>
          {{ transactionDirection }} {{ lastTransaction.amount | currency(lastTransaction.type) }}
        </v-list-tile-sub-title>
      </template>

      <!-- Message -->
      <template v-else>
        <v-list-tile-sub-title
          v-if="readOnly || isMessageI18n"
          v-text="isMessageI18n ? $t(lastMessageText) : lastMessageText"
        />
        <v-list-tile-sub-title v-else>{{ lastMessageTextNoFormats }}</v-list-tile-sub-title>
      </template>
    </v-list-tile-content>

    <div class="chat-preview__date">
      {{ createdAt | date }}
    </div>
  </v-list-tile>
</template>

<script>
import { removeFormats } from '@/lib/markdown'

import transaction from '@/mixins/transaction'
import dateFilter from '@/filters/date'
import ChatAvatar from '@/components/Chat/ChatAvatar'
import Icon from '@/components/icons/BaseIcon'
import AdmFillIcon from '@/components/icons/AdmFill'

export default {
  mounted () {
    // fetch status if last message is transaction
    if (this.lastTransaction) {
      this.fetchTransactionStatus(this.lastTransaction, this.partnerId)
    }
  },
  watch: {
    // fetch status when new message received
    lastTransaction () {
      this.fetchTransactionStatus(this.lastTransaction, this.partnerId)
    }
  },
  computed: {
    userId () {
      return this.$store.state.address
    },
    partnerName () {
      return this.$store.getters['partners/displayName'](this.partnerId) || this.partnerId
    },
    lastMessage () {
      return this.$store.getters['chat/lastMessage'](this.partnerId)
    },
    isMessageI18n () {
      return this.lastMessage.i18n
    },
    lastMessageText () {
      return this.$store.getters['chat/lastMessageText'](this.partnerId)
    },
    lastMessageTextNoFormats () {
      if (
        this.readOnly ||
        this.$store.state.options.formatMessages
      ) {
        return removeFormats(this.lastMessageText)
      }

      return this.lastMessageText
    },
    lastMessageTimestamp () {
      return this.$store.getters['chat/lastMessageTimestamp'](this.partnerId)
    },
    lastTransaction () {
      if (this.lastMessage) {
        const abstract = this.lastMessage

        if (abstract.type !== 'message') {
          return abstract
        }
      }

      return null
    },
    transactionDirection () {
      const direction = this.userId === this.lastTransaction.senderId
        ? this.$t('chats.sent_label')
        : this.$t('chats.received_label')

      return direction
    },
    numOfNewMessages () {
      return this.$store.getters['chat/numOfNewMessages'](this.partnerId)
    },
    createdAt () {
      return this.lastMessageTimestamp
    },
    status () {
      return this.getTransactionStatus(this.lastMessage)
    },
    statusIcon () {
      if (this.status === 'delivered') {
        return 'mdi-check'
      } else if (this.status === 'pending') {
        return 'mdi-clock-outline'
      } else if (this.status === 'rejected') {
        return 'mdi-close-circle-outline'
      } else {
        return 'mdi-alert-outline'
      }
    }
  },
  data: () => ({
  }),
  filters: {
    date: dateFilter
  },
  mixins: [transaction],
  components: {
    ChatAvatar,
    Icon,
    AdmFillIcon
  },
  props: {
    partnerId: {
      type: String,
      required: true
    },
    readOnly: {
      type: Boolean,
      default: false
    }
  }
}
</script>

<style lang="stylus" scoped>
@import '~vuetify/src/stylus/settings/_colors.styl'

.chat-preview
  position: relative

  &__date
    font-size: 8px
    font-style: italic
    color: $grey.base
    position: absolute
    top: 16px
    right: 16px

/** Themes **/
.theme--light
  .chat-preview__icon
    background-color: $grey.lighten-1
    color: $shades.white
  .adm-icon
    fill: #BDBDBD

.theme--dark
  .chat-preview__icon
    background-color: $grey.darken-1
</style>
