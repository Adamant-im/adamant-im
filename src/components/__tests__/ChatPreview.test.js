import { shallowMount } from '@vue/test-utils'
import Vue from 'vue'
import Vuex from 'vuex'
import VueI18n from 'vue-i18n'
import Vuetify from 'vuetify'

import formatters from '@/lib/formatters'
import mockupI18n from './__mocks__/plugins/i18n'
import ChatPreview from '@/components/ChatPreview'

Vue.use(Vuetify)
Vue.use(VueI18n)
Vue.use(Vuex)

// set $formatAmount & $formatDate prototype methods
formatters(Vue)

/**
 * Fake vars for testing getters.
 */
let fake = createFakeVars()

function createFakeVars () {
  return {
    partnerId: 'U123456',
    partnerName: 'Rick',
    lastMessage: {
      id: 1,
      type: 'message',
      message: 'you are death if you touch me'
    },
    lastMessageText: 'you are death if you touch me',
    lastMessageTimestamp: 0, // 01.01.1970
    numOfNewMessages: 0
  }
}

/**
 * Mockup store helper
 */
function mockupStore () {
  const partners = {
    getters: {
      displayName: state => partnerId => fake.partnerName
    },
    namespaced: true
  }

  const chat = {
    getters: {
      numOfNewMessages: state => partnerId => fake.numOfNewMessages,
      lastMessage: state => partnerId => fake.lastMessage,
      lastMessageText: state => partnerId => fake.lastMessageText,
      lastMessageTimestamp: state => partnerId => fake.lastMessageTimestamp
    },
    namespaced: true
  }

  const language = {
    state: {
      currentLocale: 'en'
    },
    namespaced: true
  }

  const store = new Vuex.Store({
    modules: {
      partners,
      chat,
      language
    }
  })

  return {
    store,
    partners,
    chat,
    language
  }
}

describe('ChatPreview.vue', () => {
  let i18n = null
  let store = null
  let chat = null
  let partners = null
  let language = null

  beforeEach(() => {
    const vuex = mockupStore()
    store = vuex.store
    chat = vuex.chat
    partners = vuex.partners
    language = vuex.language

    i18n = mockupI18n()
    fake = createFakeVars() // reset fake vars
  })

  it('renders the correct markup', () => {
    const wrapper = shallowMount(ChatPreview, {
      store,
      i18n,
      propsData: {
        partnerId: fake.partnerId
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('check computed properties', () => {
    fake.partnerId = 'U654321'
    fake.partnerName = 'Morty'
    fake.lastMessageText = 'ok'
    fake.lastMessageTimestamp = 3600 // Sep 2, 21:00
    fake.numOfNewMessages = 1

    const wrapper = shallowMount(ChatPreview, {
      store,
      i18n,
      propsData: {
        partnerId: fake.partnerId
      }
    })

    expect(wrapper.vm.partnerName).toBe('Morty')
    expect(wrapper.vm.lastMessageText).toBe('ok')
    expect(wrapper.vm.lastMessageTimestamp).toBe(3600)
    expect(wrapper.vm.numOfNewMessages).toBe(1)
    expect(wrapper.vm.createdAt).toBe(3600)
  })
})
