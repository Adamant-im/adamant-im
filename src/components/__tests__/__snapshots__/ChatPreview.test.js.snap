// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatPreview.vue renders the correct markup 1`] = `
<v-list-tile-stub
  activeclass="primary--text"
  class="chat-preview"
>
  <v-list-tile-avatar-stub
    size="40"
  >
    <chat-avatar-stub
      size="40"
      usepublickey="true"
      userid="U123456"
    />
     
    <v-badge-stub
      color="primary"
      overlap="true"
      transition="fab-transition"
      value="true"
    >
      <!---->
    </v-badge-stub>
  </v-list-tile-avatar-stub>
   
  <v-list-tile-content-stub>
    <v-list-tile-title-stub>
      Rick
    </v-list-tile-title-stub>
     
    <v-list-tile-sub-title-stub>
      you are death if you touch me
    </v-list-tile-sub-title-stub>
  </v-list-tile-content-stub>
   
  <div
    class="chat-preview__date"
  >
    
    Jan 01, 1970, 02:00
  
  </div>
</v-list-tile-stub>
`;
